Add support for holes when triangulating HeightMapShape3D for navigation

bluejay-sh opened the pull request:
This is a continuation of https://github.com/godotengine/godot/pull/101380 with additional changes as discussed in https://github.com/godotengine/godot-proposals/issues/11531, "Standardize label for holes in heightfield collider data"

This pull adds two methods to the physics server - `heightmap_shape_is_hole_label`, and `heightmap_shape_get_hole_labels`. Both methods are exposed to GDscript, and have been implemented for GodotPhysics and Jolt.

`heightmap_shape_get_hole_labels` returns a list of heightfield values that the physics engine will interpret as holes when computing collision for a heighmap shape.

`heightmap_shape_is_hole_label` is a convenience function that will return true if the value passed to it represents a hole.

As in the original PR, the logic used to triangulate heightmap collision shapes has been modified to support holes in the heightfield, testing points using `heightmap_shape_is_hole_label`.

![image](https://github.com/user-attachments/assets/031f24b0-b280-4121-ac4c-5be6e0452a3a)

- *Production edit: This closes https://github.com/godotengine/godot-proposals/issues/2746 and closes https://github.com/godotengine/godot-proposals/issues/11531.*

bluejay-sh committed:
Initial

bluejay-sh committed:
Merge branch 'godotengine:master' into landscape-hole-support

smix8 commented:
See https://docs.godotengine.org/en/stable/contributing/documentation/updating_the_class_reference.html#updating-class-reference-when-working-on-the-engine
You need to run `--doctool` and fix the missing XML documentation for CI to pass.

bluejay-sh committed:
Update documentation and add argument name to `heightmap_shape_is_hole_label` binding.

bluejay-sh committed:
...

bluejay-sh commented:
Yeah, just caught that, and realized that I hadn't assigned an argument name to the is_label binding call. This should do it.

mihe commented:
Without having tried this PR myself, and not knowing much about the navigation side of Godot, from where exactly does `StaticBody3D::navmesh_parse_source_geometry` get called?

When running with `physics/3d/run_on_separate_thread` enabled, a `FUNC1RC` method will block until the main thread synchronizes with the physics thread, so could be problematic if called at the wrong time/place.

The `heightmap_shape_is_hole_label` function in general strikes me as a bit overkill. You're not passing in the shape `RID` together with the height value, so I can't imagine there's any relevant context that will change what height values represent a hole during runtime. Not that there really should be, even if you did pass in the shape. I feel like it might better to only expose something like `heightmap_shape_get_hole_labels` and have it not be a synchronizing function, meaning don't use the `FUNC*` macros in `PhysicsServer3DWrapMT`.

I would perhaps question the naming as well. Using the prefix of `heightmap_shape_*` would suggest that it's a "method" of the `RID` belonging to a heightmap shape, but you're not passing in any such `RID`, so looks a bit off I would say. On top of that, I don't think I've seen this referred to as "labels" anywhere else in Godot.

Maybe calling it `PhysicsServer3D.get_heightmap_hole_constants` is more appropriate?

But even then, I'm not completely sold on this needing to support multiple constants. The `FLT_MAX` constant that Jolt uses is arguably an implementation detail, and while you can technically supply that right now, it's not really documented anywhere outside of Jolt's documentation, and for all intents and purposes `NAN` (which I don't believe is portable, unfortunately) should be the only constant of any relevance.

So I would probably suggest just going with `PhysicsServer3D.get_heightmap_hole_constant` instead, and have it return `NAN` for both Godot Physics and Jolt.

smix8 commented:
@mihe
The `navmesh_parse_source_geometry()` is a callback functions called by the `NavMeshGenerator3D` for every single register `NavMeshGeometryParser3D` parser and every `Node` parsed in the navmesh source geometry parsing process.

At the moment this parsing is entirely locked to the main thread as the `SceneTree` is not thread-safe and blocks half the nodes and their parsing when called from a thread.

A parsing can be triggered anytime by users in script when they call the `NavigationServer3D.parse_source_geometry_data()` function or the similar function on a `NavigationRegion3D.bake_navigation_mesh()`.

bluejay-sh commented:
@mihe 
I do prefer the `get_heightmap_hole_constants` nomenclature to what's currently in the PR. I'll be sure to update the names to reflect this when I go to pull out the FUNC macros. I had initially anticipated adding the functions to HeightMapShap3D but held off until I could get feedback on the PR. Given the role they play it's probably better to keep them at the level of the physics engine-proper.

I think that multiple returns are the way to go because it's entirely possible for a physics engine to not support holes in its heightfield to begin with, and even though in practice nobody is likely to accidentally run into collision issues using a value like FLT_MAX in practice, it's better to err on the side of informing the user ahead of time.

If we did go in the direction of returning a single constant I believe it might be best to return the native no-collision value unless the engine performs a data transformation to prevent it from registering as a hole in the in the first place. Between NaN and FLT_MAX, NaN is the value a user would most likely expect to exhibit special or undocumented behavior since it's somewhat magical by default.

`is_hole_label` saves code that needs to detect holes from having to account for the non-equivalency of NaN with itself, which might not even be an issue to consider in the case of some implementations. It's less than ideal, but I think it's necessary.

mihe commented:
> having to account for the non-equivalency of NaN with itself

Right, I forgot about that unfortunate detail.

I'd be curious to know what the performance overhead is of calling a virtual method 6 times for every quad in the heightmap, and whether it stands out in this particular context compared to just calling `Math::is_nan` directly.

> I think that multiple returns are the way to go because it's entirely possible for a physics engine to not support holes in its heightfield to begin with and even though in practice nobody is likely to accidentally run into collision issues using a value like FLT_MAX in practice, it's better to err on the side of informing the user ahead of time.

> If we did go in the direction of returning a single constant I believe it might be best to return the native no-collision value unless the engine performs a data transformation to prevent it from registering as a hole in the in the first place

I struggle to see the need for a potential native/redundant constant to be a runtime queryable thing, especially in this form.

I can't see any physics engine having a magic constant that could also be a reasonable height value in other engines, which only leaves the case of returning no constant to indicate that holes aren't supported, which I feel could be better served by something like `is_heightmap_holes_supported`. But at that point I would probably rename it to `is_feature_supported` and have it take some `FEATURE_HEIGHTMAP_HOLES` enum instead. There's probably a few other things we could add to that enum that would prove useful.

However, given the Godot mantra of only adding things that have an actual practical need for someone in the moment, and with it not being what's actually needed here (if we assume the presence of some `is_heightmap_hole(float p_height)` function instead) I would probably hold off on adding it altogether for now.

---

Frankly, the more I think about all this, the more I'm inclined to say that we should just properly document `NAN` as the hole constant across the board and leave it at that. Anyone integrating a physics engine just needs to do the transformation on their end, and accept that we'll have potentially redundant constants, like with Jolt.

While `NAN` isn't the most ergonomic perhaps, it should be immediately obvious that something isn't working right if someone does try to do an equality comparison, as opposed to something like `std::is_nan` (or equivalent in other languages), so doesn't strike me as a serious problem in practice.

bluejay-sh commented:
> Frankly, the more I think about all this, the more I'm inclined to say that we should just properly document `NAN` as the hole constant across the board and leave it at that. Anyone integrating a physics engine just needs to do the transformation on their end, and accept that we'll have potentially redundant constants, like with Jolt.

If standardizing around NaN is still on the table, then I think you're right about pulling `heightmap_shape_get_hole_labels`. I'll also leave `is_heightmap_hole` completely internal instead of exposing it to scripting. It only makes sense to expose publicly if the engine doesn't explicitly guarantee a particular hole constant, and it might later be beneficial to strip the function out entirely and update the triangulation code to use NaN checks after checking for feature support.

bluejay-sh committed:
- remove `heightmap_shape_get_hole_labels`
- rename `heightmap_shape_is_hole_label` to `is_heightmap_hole`
- `is_heightmap_hole` is no longer exposed to gdscript
- `is_heightmap_hole` is no longer wrapped with FUNC1RC in `physics_server_3d_wrap_mt.h`

This PR will no longer close https://github.com/godotengine/godot-proposals/issues/11531.

smix8 reviewed (dismissed):
For the navigation part.

mihe reviewed (dismissed):
Oh, sorry, this ended up falling through the cracks, my bad.

As discussed already, the current exposed/virtual `is_heightmap_hole` method doesn't make much sense in its current form if we're committing to only NaN as the hole constant. It's also in a bit of an inbetween state right now, due to exposing the method to extensions but not scripts.

So this needs to be changed to either:

1. Have `is_heightmap_hole` be a non-virtual inlined method in `PhysicsServer3D` that just calls `Math::is_nan`.
2. Omit `is_heightmap_hole` entirely and rely on `Math::is_nan` in `navmesh_parse_source_geometry`.

I'm in favor of _only_ documenting NaN and leaving it that, meaning omitting `PhysicsServer3D::is_heightmap_hole` entirely and relying on `Math::is_nan` in `navmesh_parse_source_geometry` instead.

If we have `PhysicsServer3D::is_heightmap_hole` as purely an internal helper method, it seems like it'd just be a matter of time before someone asks to expose it to scripts/extensions, and arguing against that when we're using it internally would seem a bit odd, so I feel like we might as well not go down that route.

mihe reviewed (changes requested):
Since this PR seems to have gotten stuck in a bit of a limbo, and in the interest of getting this merged, I'll go ahead and make the path forward less ambiguous.

@bluejay-sh Please revert all the physics server changes and have `StaticBody3D::navmesh_parse_source_geometry` use `Math::is_nan` in place of the current `physics_server->is_heightmap_hole` calls.

Any documentation changes or helper methods can be done in a separate/future PR.

NaN is already more or less the official (albeit undocumented) hole constant for Godot, and relied upon by popular extensions like Terrain3D, so the reliance upon this in `navmesh_parse_source_geometry` wouldn't really change much in this regard.

smix8 commented:
Just (re)confirming was mihe said. We had a discussion about this so let's add the NaN version to the navmesh baking part and leave the PhysicsServer part out and call it a day for now.
